modules/plugin/shapefile/src/test/java/org/geotools/data/shapefile/ShapefileDataStoreTest.java

```diff
@@ -883,6 +883,68 @@ public void testUpdating() throws Throwable {
         sds.dispose();
     }
 
+    @Test
+    public void testUpdateMultipleAttributesNoAutocommit() throws Exception {
+        // create feature type
+        SimpleFeatureType type =
+                DataUtilities.createType(
+                        "junk", "the_geom:Point,b:java.lang.Integer,c:java.lang.Integer");
+        DefaultFeatureCollection features = new DefaultFeatureCollection();
+
+        SimpleFeatureBuilder build = new SimpleFeatureBuilder(type);
+        SimpleFeature feature = null;
+        for (int i = 0; i < 3; i++) {
+            build.add(new GeometryFactory().createPoint(new Coordinate(i, i)));
+            build.add(i);
+            feature = build.buildFeature(null);
+            features.add(feature);
+        }
+
+        // store features
+        File tmpFile = getTempFile();
+        tmpFile.createNewFile();
+        ShapefileDataStore s = new ShapefileDataStore(tmpFile.toURI().toURL());
+        writeFeatures(s, features);
+
+        // read them back
+        FeatureReader<SimpleFeatureType, SimpleFeature> reader = s.getFeatureReader();
+        // System.out.println(DataUtilities.list(features));
+        reader.close();
+
+        Transaction transaction = new DefaultTransaction();
+        SimpleFeatureStore store =
+                (SimpleFeatureStore) s.getFeatureSource(s.getSchema().getTypeName(), transaction);
+
+        FilterFactory2 ff = CommonFactoryFinder.getFilterFactory2(null);
+        Query query = new Query(s.getSchema().getTypeName());
+        for (int i = 0; i < 3; i++) {
+            query.setFilter(ff.equal(ff.property("b"), ff.literal(i), true));
```

Comment:
This one modifies the b matching i, changing them to -i. And sets c (for whatever reason, is it related to the bug report?).
I think I see this code eventually calling internally getReader with the transaction set, so it will end up setting all features with the current value of -i, instead of just the desired one. Eventually it proves the fix, but there's quite some distance between what the report says and what the test ends up doing (which is, the way I see it, proving that a filtered call to modify changes only the desired records, and not the others).

Reply:
Hi @aaime 
I admit that the report title and the description is not very clear. 
The problem was discovered in udig while editing multiple features of a shapefile layer within the same transaction (auto commit off). There we observe that during the commit action all affected attributes of modified features were set to the value of the last modified feature. We traced back the problem and figure out that the source of it was the mentioned fragment of the ContentFeatureStore.getReader(Query query) method.
I suggest to change the PR title (and jira one, is it possible or do we need to close it and create an new jira ticket??) to something more clear based on the above explanation i.e:

- _"Modifications of multiple shapefile features within the same transaction result in erroneous data updates"_

and the jira description to the following:

- _"While editing multiple features of a shapefile layer within the same transaction (auto commit off) it was observed that during the commit action all affected attributes of modified features were set to the value of the last modified feature"._

and provide a detailed explanation of the root cause in a subsequent jira comment section:

- _"The root cause of the bug was traced to ContentFeatureStore.getReader(Query query) method. More specifically a ContentFeatureStore subclass that does not return true from ContentFeatureSource.canTransact() - e.g. ShapeFileFeatureStore - will return invalid results from ContentFeatureStore.getReader(Query query). This has the side effect that, if a transaction is open, all added and modified features will be returned by the iterator, even if they do not match the query."_

Would this be sufficient? (note that the jira ticket was created after my suggestion but another contributor to udig project)

Reply:
@nprigour yep the suggestion works thanks!